### ISSUE TITLE
Scripts arguments handling of multiple UserInput/selections refactoring 

### ISSUE THREAD

**Issue by jlovs** [gitextensions/gitextensions#4395]
Feature Request

Each user input/selection when configured in Scripts Arguments will show a separate dialog.
There is no way of telling the user of a script what ${UserInput} or ${sBranch} means.

Suggestion

Let an argument variable have the possibility to have a label {argument:label}
"{UserInput:CommitMessage}" {sBranch:MasterBranch}
Create a ScriptInputDialog that provide a dynamic input view with all arguments for a script in one place.
And make it possible to cancel the script from that dialog.
The citation around {UserInput} would also be nice to have, to make it possible to pass strings with spaces inside.
Make it possible to execution the script inside a console that can handle input. "Run in Console" checkbox?
And finally there is no argument that gives all the existing branches, or did I miss something?

Why?
To make it possible to call scripts that need many parameters and understand what to provide to it.
I personally need it to handle submodules in a nice and easy way.

If help is needed I would be glad to contribute.

**Comment by vbjay** [comment]
Submodule scripts are handled like this.
![image](https://user-images.githubusercontent.com/3629489/35449418-f03f4cb2-028a-11e8-99d6-a414467574e1.png)

**Comment by jlovs** [comment]
Yes I know that, it is not the problem for me. I have a set of scripts that I want to run from a GUI tool. And my needs are more than git extensions can provide right now.

**Comment by gerhardol** [comment]
> If help is needed I would be glad to contribute.

I can put it this way: There are a lot more users than contributors and maintainers.

Seems like a reasonable feature

**Comment by jlovs** [comment]
Ok, I'll put some effort into it this weekend and probably get back with some questions and suggestions for review.

I use VS2017 today and I guess that is out of the question because of the .csproj .sln files changing, or is it possible to develop in that version and provide a pull-request from that?

I had to change some things to make it build and compile a release, but not that much.

**Comment by RussKie** [comment]
I use VS2017 fulltime.
We're also planning to fully migrate to it in the next release.

> I'll put some effort into it this weekend

If I may suggest before you spend too much time coding it, create some dummy screens to show how you envisage the feature, and provide a feel for the proposed flows.
Thanks

**Comment by jlovs** [comment]
Yes of course. To be honest I have a draft already locally that I'm happy with regarding the function of it. Just have to make it more of a production code and add the dynamic dialog for the argument option input.

But I will, as you suggest provide some dummy screens to let you get a feeling of my thoughts and hopefully provide me with some feedback and suggestions.

I'll get back soon.

**Comment by gerhardol** [comment]
Yes, you can use MSVS2017

#4311 is for setting 2017 as default, planned for the release after 2.51
I have cleaned up in the build setup to simplify builds. Issue #4269/PR#4270 handles dependencies too. I have done updates to buildscripts too that aligns the remaining 2015/2017 differences too, no PR yet. I consider dropping #4270 and make all changes in one PR.

**Comment by jlovs** [comment]
Something like this:
Add an option to select Run in console for a script, to run it in a console that can take input.
In the help showed when configuring script "tags", there will be an explanation of how to provide a label for a "tag". Something like {tag:label}.
When the script is to be run, loop through all arguments with Split(' ') and collect all arguments with  a ':'.
Show a dialog that takes all those arguments as parameters, that shows a label and input per argument.
If argument has selectable values a ComboBox is used, otherwise a TextBox, aside a label.
When dialog is closed the arguments typed in/selected will replace the "tags".

After that the normal functionality for "tags" will be called and used as today. Or it is made mandatory to use a label for a "tag".

Add some new tags {aRefs} {aBranches} {aTags}

Run in console will use the ConsoleEmulatorOutputControl instead of the EditboxBasedConsoleOutputControl.

Some screens:
![settings-full](https://user-images.githubusercontent.com/24225477/35485332-760db64c-045e-11e8-81bb-37536bf6f6d6.PNG)
![setarguments](https://user-images.githubusercontent.com/24225477/35485342-936e5610-045e-11e8-9573-78fbf36916e6.PNG)
![console](https://user-images.githubusercontent.com/24225477/35485399-87edaf2e-045f-11e8-8cd2-357b59bde8ff.PNG)

**Comment by vbjay** [comment]
Did you look at the git flow support we have?

  http://git-extensions-documentation.readthedocs.io/en/latest/plugins.html

**Comment by jlovs** [comment]
@vbjay Git flow is something that I have looked into before. The problem is that it does not take care of submodules in an automatic way, and the branching strategy is far to complex for our needs. But the idea is good to isolate development in feature-ish branches.

**Comment by jlovs** [comment]
Should I provide a PR for review?

Or is it better to discuss and specify more first?

What do you think?

**Comment by RussKie** [comment]
If you have already something to show - please send it in.

I haven't had a chance to give the proposal a deep thought (too many other
things going on), but the screenshots somewhat scare me...

I also can't promise you a quick turnaround response either - real life and
the existing backlog all demanding attention.

On 30/01/2018 5:11 PM, "jlovs" <notifications@github.com> wrote:

Should I provide a PR for review?

Or is it better to discuss and specify more first?

What do you think?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/gitextensions/gitextensions/issues/4395#issuecomment-361487624>,
or mute the thread

**Comment by jlovs** [comment]
@RussKie Well tell me about real life ;) Got two small kids half-time, full-time job, long distance runner...

Scare you? Please provide me with some more info about what scares you!

I've been coding some yes, but that is not production code yet, because of some outstanding questions regarding:
- Translation
- Keep old functionality or not
- Label format (no spaces, convert label config to leading uppercase or not)
- Where to put stuff (can be reviewed of course)
- VS2017 csproj files that I have updated. How will they be merged, when there is no 2017 .sln file yet?
- Adding new icons outside of installation to scripts.
- And probably more when coding it...

Finally I respect that you have other things to do, but hope this can be brought forward in some way.

**Comment by RussKie** [comment]
Heh, I'm very much in the same boat - +kids, +work, -running,
+gitextensions.
I generally get a few dozens of emails a day from github... You, users, are
a demanding bunch :) ...which is good in a way.

>  Please provide me with some more info about what scares you!

I will get back to you when I have time to review the proposal in more
details.

> Translations

These are handled automagically via Transifex. You need to use
`TranslationString` construct or set strings in .Desginer.cs file.

> Keep old functionality or not

If there is no regression and the new functionality provides parity, it is
cleaner, meaner, SOLIDer and testable - then I'm all for the new
functionality :)

>  Label format

Follow the existing styling

> VS2017

I use VS2017 exclusively, so no dramas here. We are planning to fully move
to 2017 in a near future.

>  Adding new icons outside of installation to scripts

Please elaborate.

**Comment by jlovs** [comment]
Thanx @RussKie @gerhardol 
I get back as soon as I get some spare time and put a link with a preview.
Delivery time at work, and race on sunday.

Icons... Just point out an icon somewhere on the local harddrive.

Have a nice weekend!
Btw... Stockholm Marathon in June? I just decided to run :1st_place_medal:

**Comment by jlovs** [comment]
> Will probably be a pacer at 3:30!

Cool, I will hopefully get to the goal 15 min before you then, or at least 10.

**Comment by LukeOwlclaw** [comment]
@jlovs Did you make any more progress with this feature? I do not see a corresponding branch in your cloned GE repo...

**Comment by GintasS** [comment]
Abandoning this because of age and lack of discussion.

**Comment by h0lg** [comment]
+1 Yes, please! Multiple named user inputs for routine and maintenance scripts would be a charm <3.

Also, please consider the ability of chaining commands.

Currently, I'm making due with aliases like these in my `.gitconfig`:

```
[alias]
	rebaseonmaster = !sh -c 'git checkout \"$0\" && git rebase \"master\" && git checkout \"$1\"'
	renameBranchRemotely = !sh -c 'git branch -m \"$1\" && git push origin --delete \"$0\" && git push origin -u \"$1\"'
```

Please, have mercy on my eyes!
